aspnet/fundamentals/app-state.rst

```diff
@@ -1,8 +1,252 @@
+
+Other Persistence
+^^^^^^^^^^^^^^^^^
+Any other form of persistent storage, whether using Entity Framework and a database or something like Azure Table Storage, can also be used to store application state, but these fall outside of what ASP.NET supports directly.
+
+Working with HttpContext.Items
+------------------------------
+The ``IHttpContext`` abstraction provides support for a simple object ``NameValueCollection``, called ``Items``. This collection is available from the start of an `IHttpRequest`` and is disposed of at the end of each request. You can access it by simply assigning a value to a keyed entry, or by requesting the value for a given key.
+
+For example, some simple :doc:`middleware` could add something to the ``Items`` collection:
+
+.. code-block:: c#
+
+	app.Use(async (context, next) =>
+		{
+			// perform some verification
+			context.Items["isVerified"] = true;
+			await next.Invoke();
+		});
+
+and later in the pipeline, another piece of middleware could access it:
+
+.. code-block:: c#
+
+	app.Run(async (context) =>
+	{
+		await context.Response.WriteAsync("Verified request? "
+			+ context.Items["isVerified"]);
+	});
+
+.. note:: Since keys into ``Items`` are simple strings, if you are developing middleware that needs to work across many applications, you may wish to prefix your keys with a unique identifier to avoid key collisions (e.g. "MyComponent.isVerified" instead of just "isVerified").
+
+Installing and Configuring Session
+----------------------------------
+
+ASP.NET 5 ships a session package that provides middleware for managing session state. You can install it by including a reference to the package in your project.json file:
+
+.. literalinclude:: app-state/sample/src/AppState/project.json
+	:language: javascript
+	:linenos:
+	:lines: 5-9
+	:emphasize-lines: 4
+
+Once the package is installed, Session must be configured in your application's ``Startup`` class. Session is built on top of ``IDistributedCache``, so you must configure this as well, otherwise you will receive an error.
+
+.. note:: If you do not configure at least one ``DistributedCache`` implementation, you will get an exception stating "Unable to resolve service for type 'Microsoft.Framework.Caching.Distributed.IDistributedCache' while attempting to activate 'Microsoft.AspNet.Session.DistributedSessionStore'."
+
+ASP.NET ships with several implementations of ``IDistributedCache``, including an in-memory option (to be used during development and testing only). To configure session using this in-memory option, add the following to ``ConfigureServices``:
+
+.. code-block:: c#
+
+	services.AddCaching();
+	services.AddSession();
+
+Then, add the following to ``Configure`` and you're ready to use session in your application code:
+
+.. code-block:: c#
+
+	app.UseSession();
+
+You can reference Session from ``IHttpContext`` once it is installed and configured.
```

Comment:
Check the kestrel logs. Because the exception happened after the response started it can't be gracefully handled and rendered as an error page. We log it and reset the connection.